python/ray/data/dataset.py

```diff
@@ -2357,11 +2365,19 @@ def take(self, limit: int = 20) -> List[Dict[str, Any]]:
                 "records in pandas or numpy batch format."
             )
         output = []
-        for row in self.iter_rows():
+        from ray.data.dataset_pipeline import DatasetPipeline
+
+        is_not_pipeline = not isinstance(self, DatasetPipeline)
+        limited_ds = self.limit(limit) if is_not_pipeline else self
+        for row in limited_ds.iter_rows():
             output.append(row)
             if len(output) >= limit:
                 break
         self._synchronize_progress_bar()
+        if is_not_pipeline:
+            # Save the computed stats and blocks to the original dataset.
+            self._plan._snapshot_stats = limited_ds._plan.stats()
+            self._plan._snapshot_blocks = limited_ds._plan._snapshot_blocks
```

Comment:
can we copy this method into `DatasetPipeline`, similar to how we have done for `iter_torch_batches` for example? 

`DatasetPipeline` no longer needs to be maintained, so just snapshotting the current implementation and copying it over would suffice instead of leaking `DatasetPipeline` details into `Dataset`.

Reply:
I think what Amog suggests is that you can copy the old code without limit to the DatasetPipeline. Then we don't need these additional checks here.

Reply:
yep exactly, we can just copy the old code.

Reply:
Separately, can we add a deprecation warning of `DatasetPipeline`?

Reply:
@c21 do you mean we should add a deprecation warning for using `DatasetPipeline` in a separate PR?

Reply:
yes, in a separate PR.